src/Control/Monad/Catch/Pure.hs

```diff
@@ -159,6 +159,17 @@ instance Monad m => MonadCatch (CatchT m) where
 instance Monad m => MonadMask (CatchT m) where
   mask a = a id
   uninterruptibleMask a = a id
+  generalBracket acquire release cleanup use = CatchT $ do
+    eresource <- runCatchT acquire
+    case eresource of
+      Left e -> return $ Left e
+      Right resource -> do
+        eresult <- runCatchT (use resource)
+        case eresult of
+          Left e -> do
+            _ <- runCatchT (cleanup resource e)
+            return $ Left e
+          Right result -> runCatchT (release resource result)
```

Comment:
This `CatchT` instance (which I would expect to be more-or-less the same as the `Either` instance) appears to be subtly different from the `Either` instance. In the `CatchT` instance, the returned value when there are no exceptions is the result of `release`. But in the `Either` instance, you have:

```haskell
Right result -> release resource result >> return result
```

Is this intentional?